### DISCUSSION TITLE
mpfs-rpmsg-bm(remote) "start_demo" affected timer interrupt print on uart

### DISCUSSION THREAD

**Discussion by EricZhangjie** [polarfire-soc/polarfire-soc-documentation#144] [Applications and demos] [answered]
hello，I am running linux+BM demo on icicle kit，the BM program used mpfs-rpmsg-bm(remote) ,config linux uboot running on u54_1、u54_2、u54_3, and BM running on u54_4. I add a Timer interrupt and print a message per second on uart3, also some messages when timer is enabled. The red lines below are added lines.
![1658308227143_7BA51603-AC4C-4618-B16A-59CA76612E7B](https://user-images.githubusercontent.com/109322517/179947008-bdb564da-e947-4bad-b64c-d466631ab8a9.png)
![1658308248418_F3F09C59-8C09-4494-B274-27861945EDC9](https://user-images.githubusercontent.com/109322517/179947071-27e475cd-3553-4671-8aa3-c7ff2f013ac3.png)
![1658308267658_6759EF34-F3E3-4203-98E3-707F47B8536E](https://user-images.githubusercontent.com/109322517/179947102-921ba82d-d06c-41e8-8e08-bdaea7f1569a.png)
The uart3 print result:
![1658308011055_DB4C2DE1-000B-4e5e-951A-98A97438FCD0](https://user-images.githubusercontent.com/109322517/179948496-59eca4f4-fb7f-47ca-b43d-103fac55a8bd.png)
As BM code only running on hart4, so I modified makefile and do not compile hart0,hart1,hart2,hart3 codes ins application folder, then I find uart3 do not print "[Dsp] timer_cnt = ..."message per second in timer interrupt.
![image](https://user-images.githubusercontent.com/109322517/179950752-e0f914d1-3d5c-448f-a908-a0c44ec280ff.png)
![1658307765544_041EE464-FA16-45a1-B243-6C5F1E22A32F](https://user-images.githubusercontent.com/109322517/179948752-f4c12bef-36ed-4f39-bd7b-d3697137bf8f.png)
I find "start_demo->rpmsg_setup"function  in u54_1 that affected the result. It does not print timer interrupt message if remove "start_demo" in u54_1. I modified "start_demo",and only "rpmsg_setup" left.
![1658307803728_3AD1A749-B2D8-4969-8181-10EC9AD063BA](https://user-images.githubusercontent.com/109322517/179949585-3c020a27-4a32-42a0-b046-ca167c72cee6.png)
![1658307848534_FF1302A7-3A42-4ada-8A03-A2E500105A84](https://user-images.githubusercontent.com/109322517/179949606-98a42529-666b-449a-9648-5c192717cce4.png)
In addition, I find if do not compile hart0,hart2,hart3 codes ins application folder, that will not affect timer interrupt print on uart3.
It is strange. The BM program runs on hart4, why u54_1 function affected?  does hart1's u54_1 function also run on any hart? how can it affected timer interrupt running?

**Comment by vfalanis** [comment]
Hi,

Have you tried debugging with GDB to find out which instruction is running on U54_4 when you see the issue? 
You can attach to GDB from SoftConsole by doing a right click on the " mpfs-rpmsg-bm-hw one-hart attach remote.launch" -> Debug as -> mpfs-rpmsg-bm hw one-hart attach remote

![image](https://user-images.githubusercontent.com/5502469/180227377-a3796ed9-8d0b-4eb8-aeac-cbfdf0ead48c.png)

Then on the debug window you can see the current program counter value (pc) and the function being run 

Also, could you please let us know which SoftConsole version are you using?

Thanks

**Reply by EricZhangjie** [reply]
My SoftConsole version is v2021.1 and v2021.3-7.0.0.599, they all have the timer interrupt print issue.
I try to debug with SoftConsole v2021.3-7.0.0.599 as you suggested, but failed, log as below
![fe58ec00929bdef9766ddb17c263c7a](https://user-images.githubusercontent.com/109322517/180401864-4c7442d2-f1af-4044-b821-3f30571a1be3.png)
![ec2b1b04848e36b27b6242e57918d83](https://user-images.githubusercontent.com/109322517/180401894-a2362d01-c8dd-470b-a8da-dc3311a3b649.png)

**Reply by vfalanis** [reply]
Hi,

If using SoftConsole in Windows, could you please try to edit the launch configuration as suggested below and try again?

1. Right click in the "mpfs-rpmsg-bm hw one-hart attach remote.launch" configuration and select Debug- as -> Debug Configurations...
![image](https://user-images.githubusercontent.com/5502469/180432403-208532bc-7fb5-439a-8be9-e27e43d6937d.png)

2. Select the "mpfs-rpmsg-bm hw one-hart attach remote" configuration in the right panel. Then go to the debugger tab and delete the following line from the "Commands" field:

file ${workspace_loc:/mpfs-rpmsg-bm/Remote-Default}/mpfs-rpmsg-remote.elf

![image](https://user-images.githubusercontent.com/5502469/180432945-90869007-7994-40aa-820a-7c2eb4af0ae5.png)

3. After deleting the line on step 2, the configuration should look as shown below:
![image](https://user-images.githubusercontent.com/5502469/180433313-a5ebf685-55ad-4e08-bcae-4d9bf6e0d77e.png)

4. Click the "Apply" button followed by the "Debug" button 

5. You should be able to see the program counter in the Debug window. For example:

![image](https://user-images.githubusercontent.com/5502469/180434875-984467da-2e43-4952-8ed2-91bbf01651a4.png)


Thanks

**Comment by nitindeshpande** [comment, marked as answer]
>  ![1658308227143_7BA51603-AC4C-4618-B16A-59CA76612E7B](https://user-images.githubusercontent.com/109322517/179947008-bdb564da-e947-4bad-b64c-d466631ab8a9.png) 

Could you add below line at line 32 in above snippet?
` (void) mss_config_clk_rst(MSS_PERIPH_MMUART3, (uint8_t) 1, PERIPHERAL_ON); `

This will bring the MMUART3 out of reset and enable clock to it.
It seems the U54_1 has that line and hence when you run code on U54_1, your UART3 also works.

**Reply by nitindeshpande** [reply]
Hi @EricZhangjie, were you able to get the issue resolved?
